Convert TezosKitError to enum type

keefertaylor opened the pull request:
Early on in this project I made the (stupid) choice to make errors `struct`s instead of `enum`s. This has lead to problematic code where underlying errors are always optional and are often just untyped strings. 

As TezosKit handles more complex interactions, it may be useful to return arrays of errors, typed errors or to know that there is no additional data associated with an error. 

This breaking change moves error handling to enums with associated values. In essence, this is a prefactor for #183 / #179.

simonmcl commented on TezosKit/TezosNode/Services/RPCResponseHandler.swift:
```diff
@@ -27,28 +27,18 @@ public class RPCResponseHandler {
-	}
-
-	// Check for \"status\":\"backtracked\" in the response, indicating a failed transaction being rolledback
-	if stringData.contains("\"status\":\"backtracked\"") {
```
@keefertaylor can you try take the code I had for catching backtracks in the other PR and replace this block with it. Feel free to refactor again, update, rename etc. But I can't update to the latest code until thats in it as it will break our app. Once you have that you can close my error handling PR entirely.

Also I came across another issue that mine wasn't catching either. If you use a customFeePolicy and set it too low (by accident). Because the estimation is never called it doesn't catch gas errors. When that happens you get a `.success(let opHash)` back from the tezosKit `send`, but the console will have "protoxxx.gasExhausted....." in it. Would be great to try catch that too

keefertaylor replied:
Acked on the gasExhaused. Filed #190

keefertaylor commented:
Hey @simonmcl, 

I added your logic in. I added a new error type, `operationError` with an associated value of an array of `OperationResponseInternalResultError`. I did this because we generate these errors inside `RPCResponseHandler` and this class isn't necessarily tied to transaction formation. 

I also added a TODO for myself to add a test here. It should be easy to come up with some fake data and ensure that we parse these errors correctly and that would help us prevent future regressions. If you happen to have an example of a response with one of these errors on hand, I'd be grateful if you could share. 

LMK what you think.

simonmcl commented:
@keefertaylor sure, below is the gas exhausted error when using hardcoded gas costs. The structure of the other errors is in this PR: https://github.com/keefertaylor/TezosKit/pull/183 . I replaced some of the large objects with `...` but they are not necessary for the error logic

```
[{"contents":[{"kind":"transaction","source":"tz1e4hAp7xpjekmXnYe4677ELGA3UxR79EFb","fee":"14464","counter":"627017","gas_limit":"139719","storage_limit":"7051","amount":"0","destination":"KT1PuoBCrK7bu9MP7LKZRhjXKZSwvgQkDrCN","parameters":{"entrypoint":"approve","value":{"prim":"Pair","args":[{"string":"KT1PxBTZFsBd1gUXhkTFpncDTjvrKbmTeHde"},{"int":"1"}]}},"metadata":{"balance_updates":[{"kind":"contract","contract":"tz1e4hAp7xpjekmXnYe4677ELGA3UxR79EFb","change":"-14464"},{"kind":"freezer","category":"fees","delegate":"tz1Ke2h7sDdakHJQh8WX4Z372du1KChsksyU","cycle":182,"change":"14464"}],"operation_result":{"status":"backtracked","storage":{"prim":"Pair","args":[{"int":"239"},{"int":"1000000"}]},"big_map_diff":[{"action":"update","big_map":"239","key_hash":"exprtkFpPpLkPNbxReKT1osWCfseKDUyyxb7URVBaX2Yw1N5Sq86hv","key":{"bytes":"0000ca1c5dd7f6665501b57c692f0726a1db46fd1d18"},"value":{"prim":"Pair","args":[[{"prim":"Elt","args":[{"bytes":"01a8969119375713736a2da93ffad0ae2f0178630100"},{"int":"1"}]}],{"int":"228"}]}}],"consumed_gas":"139619","storage_size":"6794"}}},{"kind":"transaction","source":"tz1e4hAp7xpjekmXnYe4677ELGA3UxR79EFb","fee":"62742","counter":"627018","gas_limit":"520845","storage_limit":"21605","amount":"0","destination":"KT1PxBTZFsBd1gUXhkTFpncDTjvrKbmTeHde","parameters":{"entrypoint":"tokenToXtz","value":{"prim":"Pair","args":[{"prim":"Pair","args":[{"prim":"Pair","args":[{"string":"tz1e4hAp7xpjekmXnYe4677ELGA3UxR79EFb"},{"string":"tz1e4hAp7xpjekmXnYe4677ELGA3UxR79EFb"}]},{"prim":"Pair","args":[{"int":"1"},{"int":"39756"}]}]},{"string":"2020-04-27T09:43:41Z"}]}},"metadata":{"balance_updates":[{"kind":"contract","contract":"tz1e4hAp7xpjekmXnYe4677ELGA3UxR79EFb","change":"-62742"},{"kind":"freezer","category":"fees","delegate":"tz1Ke2h7sDdakHJQh8WX4Z372du1KChsksyU","cycle":182,"change":"62742"}],"operation_result":{"status":"backtracked","storage":{"prim":"Pair","args":[{"int":"307"},{"prim":"Pair","args":[{"prim":"Pair","args":[{"prim":"Pair","args":[{"prim":"None"},{"prim":"None"}]},{"prim":"Pair","args":[{"int":"1580637600"},{"int":"10000000"}]}]},{"prim":"Pair","args":[{"bytes":"01a82322859105a5974ef58b465962428c567a204d00"},{"int":"1511"}]}]}]},"consumed_gas":"467921","storage_size":"15430"},"internal_operation_results":[{"kind":"transaction","source":"KT1PxBTZFsBd1gUXhkTFpncDTjvrKbmTeHde","nonce":0,"amount":"39756","destination":"tz1e4hAp7xpjekmXnYe4677ELGA3UxR79EFb","result":{"status":"backtracked","balance_updates":[{"kind":"contract","contract":"KT1PxBTZFsBd1gUXhkTFpncDTjvrKbmTeHde","change":"-39756"},{"kind":"contract","contract":"tz1e4hAp7xpjekmXnYe4677ELGA3UxR79EFb","change":"39756"}],"consumed_gas":"10207"}},{"kind":"transaction","source":"KT1PxBTZFsBd1gUXhkTFpncDTjvrKbmTeHde","nonce":1,"amount":"0","destination":"KT1PuoBCrK7bu9MP7LKZRhjXKZSwvgQkDrCN","parameters":{"entrypoint":"default","value":{"prim":"Right","args":[{"prim":"Pair","args":[{"prim":"Pair","args":[{"bytes":"0000ca1c5dd7f6665501b57c692f0726a1db46fd1d18"},{"bytes":"01a8969119375713736a2da93ffad0ae2f0178630100"}]},{"int":"1"}]}]}},"result":{"status":"failed","errors":[{"kind":"temporary","id":"proto.006-PsCARTHA.gas_exhausted.operation"}]}}]}}],"signature":"edsigtzVs2YoRGaZoDrvug9jKJz7KY7pbJymWqvG2QWxaM3SMECHHNVGEqRAAhAcBnXWBnNk7fzPVA6We1e644Qb2oZxgCR1cYy"}]
```

simonmcl commented:
Code looks good. I'll give it all a proper test in the app once its all merged. I have some transaction history code that needs to be merged in after

keefertaylor commented:
Cool. I'm going to land this code as is and then follow up on the gas exhausted stuff.
